### ISSUE TITLE
Symbols removed, same soname

### ISSUE THREAD

**Issue by cypriani** [taglib/taglib#757]
Hi there,

I've been taking care of updating taglib from 1.9 to 1.11 in Debian, and subsequently got this bug report: https://bugs.debian.org/839767
The problem is that several symbols disappeared (I can provide a list, if relevant), but the soname stayed the same. Only the minor version was incremented (1.14.0 in taglib 1.9, 1.16.0 in taglib 1.11).

Could you bump the soname in the next taglib release to solve this problem? Is it okay for me to use taglib.so.2.0.0 in Debian, or did you plan to use this name for the taglib2 branch?

Thanks for your help,
  Matteo

**Comment by TsudaKageyu** [comment]
Hi @cypriani,

It's our mistake that we removed some symbols. I think that we should fix the broken ABI by restoring them.
If you could provide us with the list of the removed symbols, we will restore them.

**Comment by cypriani** [comment]
Le lundi 17 octobre 2016, 17:05:11 EDT Tsuda Kageyu a écrit :
> It's our mistake that we removed some symbols. I think that we should fix
> the broken ABI by restoring them. If you could provide us with the list of
> the removed symbols, we will restore them.

Thanks! Here is a diff between the symbol list in the Debian package version 
1.9.1-2.2 and the current 1.11 version, so look for lines starting with a 
minus sign. A few of them are modified symbols; they will probably need to each 
be evaluated to see if they are backward-compatible or not.

Please let me know if I can provide any additional information.

**Comment by TsudaKageyu** [comment]
@cypriani,
Is ```String::WCharByteOrder``` the only symbol that actually causes an error on your side?
Almost all the removed symbols are private functions (I believe removing private functions won't break the ABI.), so I think that all we have to do is restore the static private variable ```String::WCharByteOrder```.

**Comment by cypriani** [comment]
Le mardi 18 octobre 2016, 00:44:59 EDT Tsuda Kageyu a écrit :
> @cypriani,
> Is ```String::WCharByteOrder``` the only symbol that actually causes an
> error on your side? Almost all the removed symbols are private functions (I
> believe removing private functions won't break the ABI.), so I think that
> all we have to do is restore the static private variable
> ```String::WCharByteOrder```.

Hm, if it's private, then I have to investigate more to see why this happened. 
I'll keep you posted. Thanks for your help!

**Comment by TsudaKageyu** [comment]
Even though it's weird, I have experienced a linkage error after removing a private static variable.
Anyway, I'm going to restore it ASAP.

**Comment by TsudaKageyu** [comment]
@cypriani,
I have brought back the variable. (Please check 97aaa0f) Could you try our latest code?

**Comment by cypriani** [comment]
Hi @TsudaKageyu,

First of all, thank you for being so reactive in helping me dealing with this issue, I appreciate greatly!

So I've did quite a bit of reading to understand this issue, and I'd like to share a few links.

First, what happened is explained in [this c2 wiki article](https://web.archive.org/web/20160709103354/http://c2.com/cgi/wiki?FragileBinaryInterfaceProblem) and in [this StackOverflow post](https://stackoverflow.com/questions/7699259/how-does-adding-a-private-member-variable-break-c-abi-compatibility).

The so-called Pimpl idiom, briefly presented in [this c2 wiki article](https://web.archive.org/web/20160807205059/http://c2.com/cgi/wiki?PimplIdiom) would allow you, in the future, to do the same kind of changes without breaking the ABI. [This other article](https://marcmutz.wordpress.com/translated-articles/pimp-my-pimpl/) has more in-depth information.

Last, but not least, the KDE project has a policy on [Binary Compatibility Issues With C++](https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B) that helps understand what you can and can't do if you want to maintain compatibility, and their way to use the Pimpl idiom.


> Even though it's weird, I have experienced a linkage error after removing a
> private static variable.

What I still don't understand, is that removing (only) **private static** variables is not supposed to break the ABI (according to the KDE link above), as long as they were never used by an inline function; which seems to be the case of TagLib::String::WCharByteOrder (I just looked quickly though).


> I have brought back the variable. (Please check 97aaa0f) Could you try our
> latest code?

The process to try to reproduce the bug would be to:

- install taglib 1.9.1
- build ncmpcpp 0.7.4 against taglib 1.9.1
- remove taglib 1.9.1 and install the latest code
- run ncmpcpp

I can try, but I'm not sure this is worth it, since it is now clear that most of the other symbol deletions and modifications are a problem as well. In my opinion, the only "fix" will be to bump the soname, unless you are willing to restore all the other symbols as well (which looks like a lot of work, and may not be worth it if you're close to releasing taglib 2).

Please let me know your thoughts.

**Comment by scotchi** [comment]
Just for the record, TagLib does use the Pimpl idiom and attempts to follow said KDE document, so places where things have disappeared inappropriately would be cases of us going against our intended policies.

**Comment by cypriani** [comment]
@scotchi: good to know. Sorry if I sounded like I wanted to teach lessons, that wasn't my intent at all :-)

**Comment by TsudaKageyu** [comment]
Again, it's really weird. However, it looks like removing a private static variable does affect the ABI.

I tried to run ncmpcpp with some versions of TagLib on a Debian 8.6 x64 virtual machine:
With `libtag1-dev/stable 1.9.1-2.1 amd64` from the official APT repository, I got this:
```
root@debian-512mb-sgp1-01:~# ncmpcpp
Couldn't connect to MPD (host = localhost, port = 6600): Connection refused
```
Of course, it was a success.

With TagLib v1.11:
```
root@debian-512mb-sgp1-01:~/code/taglib-1.11/taglib# ncmpcpp
ncmpcpp: symbol lookup error: ncmpcpp: undefined symbol: _ZN6TagLib6String14WCharByteOrderE
```
OK, I could reproduce the same error as yours.

With our latest code, I got:
```
root@debian-512mb-sgp1-01:~# ncmpcpp
Couldn't connect to MPD (host = localhost, port = 6600): Connection refused

root@debian-512mb-sgp1-01:~# ldd /usr/bin/ncmpcpp
        linux-vdso.so.1 (0x00007fffb95e7000)
        libmpdclient.so.2 => /usr/lib/libmpdclient.so.2 (0x00007f32cadfd000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f32cabe0000)
        libncursesw.so.5 => /lib/x86_64-linux-gnu/libncursesw.so.5 (0x00007f32ca9aa000)
        libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00007f32ca780000)
        libfftw3.so.3 => /usr/lib/x86_64-linux-gnu/libfftw3.so.3 (0x00007f32ca37f000)
        libcurl-gnutls.so.4 => /usr/lib/x86_64-linux-gnu/libcurl-gnutls.so.4 (0x00007f32ca112000)
        libtag.so.1 => /usr/lib/x86_64-linux-gnu/libtag.so.1 (0x00007f32c9e04000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f32c9af9000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f32c97f8000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f32c95e2000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f32c9237000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f32cb2b9000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f32c9033000)
        libidn.so.11 => /usr/lib/x86_64-linux-gnu/libidn.so.11 (0x00007f32c8dff000)
        librtmp.so.1 => /usr/lib/x86_64-linux-gnu/librtmp.so.1 (0x00007f32c8be1000)
        libssh2.so.1 => /usr/lib/x86_64-linux-gnu/libssh2.so.1 (0x00007f32c89b8000)
        libnettle.so.4 => /usr/lib/x86_64-linux-gnu/libnettle.so.4 (0x00007f32c8786000)
        libgnutls-deb0.so.28 => /usr/lib/x86_64-linux-gnu/libgnutls-deb0.so.28 (0x00007f32c8467000)
        libgssapi_krb5.so.2 => /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f32c821c000)
        libkrb5.so.3 => /usr/lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f32c7f48000)
        libk5crypto.so.3 => /usr/lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f32c7d17000)
        libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f32c7b13000)
        liblber-2.4.so.2 => /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2 (0x00007f32c7904000)
        libldap_r-2.4.so.2 => /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2 (0x00007f32c76b2000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f32c7497000)
        libboost_atomic.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_atomic.so.1.55.0 (0x00007f32c7295000)
        libhogweed.so.2 => /usr/lib/x86_64-linux-gnu/libhogweed.so.2 (0x00007f32c7066000)
        libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f32c6de3000)
        libgcrypt.so.20 => /lib/x86_64-linux-gnu/libgcrypt.so.20 (0x00007f32c6b01000)
        libp11-kit.so.0 => /usr/lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f32c68bb000)
        libtasn1.so.6 => /usr/lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f32c66a7000)
        libkrb5support.so.0 => /usr/lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f32c649b000)
        libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f32c6297000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f32c6080000)
        libsasl2.so.2 => /usr/lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f32c5e64000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f32c5c5c000)
        libgpg-error.so.0 => /lib/x86_64-linux-gnu/libgpg-error.so.0 (0x00007f32c5a4a000)
        libffi.so.6 => /usr/lib/x86_64-linux-gnu/libffi.so.6 (0x00007f32c5842000)

root@debian-512mb-sgp1-01:~# ls /usr/lib/x86_64-linux-gnu/libtag.*
/usr/lib/x86_64-linux-gnu/libtag.so.1  /usr/lib/x86_64-linux-gnu/libtag.so.1.16.0
```
ncmpcpp worked again with `libtag.so.1.16.0`.

**Comment by TsudaKageyu** [comment]
I'd like to wait for @cypriani's reply, but I think that we have fixed the issue of ABI breakage.
IMO, we have to make an "urgent release" (maybe v1.11.1) ASAP.

**Comment by cypriani** [comment]
@TsudaKageyu, thanks a lot for taking the time to test with ncmpcpp. I didn't reply right away because I didn't get feedback from other Debian people; my personal opinion is that we can consider that the problem is fixed for ncmpcpp, but expect that some other software was also broken. That being said, one could expect that if something else was broken in Debian, it would have been reported by now, so your releasing a new minor version is probably the best option.

I'll be waiting for taglib 1.11.1 if you plan to release it in the next week or so; otherwise I can just integrate your patch in the Debian package, just let me know.

**Comment by cypriani** [comment]
@TsudaKageyu, please also consider #760 for the next release, if you agree on fixing it.

**Comment by TsudaKageyu** [comment]
Hi @cypriani, we have released TagLib 1.11.1. http://taglib.github.io/releases/taglib-1.11.1.tar.gz
I hope it will solve the problems. Thanks!

**Comment by cypriani** [comment]
I uploaded taglib 1.11.1 in Debian last night. Thanks again for your help!
